src/backend/distributed/planner/recursive_planning.c

```diff
@@ -599,6 +640,337 @@ RecursivelyPlanNonColocatedSubqueriesInWhere(Query *query,
+							   nodeTag(node))));
+	}
+
+	RangeTblRef *rangeTableRef = (RangeTblRef *) node;
+	if (IsRTERefRecurring(rangeTableRef, query))
+	{
+		return;
+	}
+
+	RangeTblEntry *distributedRte = rt_fetch(rangeTableRef->rtindex,
+											 query->rtable);
+	if (distributedRte->rtekind == RTE_RELATION)
+	{
+		ereport(DEBUG1, (errmsg("recursively planning distributed relation %s "
+								"since it is part of a distributed join node "
+								"that is outer joined with a recurring rel",
+								GetRelationNameAndAliasName(distributedRte))));
+
+		PlannerRestrictionContext *restrictionContext =
+			GetPlannerRestrictionContext(recursivePlanningContext);
+		List *requiredAttributes =
+			RequiredAttrNumbersForRelation(distributedRte, restrictionContext);
+
+		ReplaceRTERelationWithRteSubquery(distributedRte, requiredAttributes,
+										  recursivePlanningContext);
+	}
+	else if (distributedRte->rtekind == RTE_SUBQUERY)
+	{
+		/*
+		 * We don't try logging the subquery here because RecursivelyPlanSubquery
+		 * will anyway do so if the query doesn't reference the outer query.
+		 */
+		ereport(DEBUG1, (errmsg("recursively planning the distributed subquery "
+								"since it is part of a distributed join node "
+								"that is outer joined with a recurring rel")));
+
+		/*
+		 * This was our only chance to plan given distributed subquery --known
+		 * to be part of the inner side of an outer join with a recurring rel--,
+		 * so we throw a hard error instead of a DeferredError if we can't
+		 * recursively plan given subquery.
+		 */
+		if (!RecursivelyPlanSubquery(distributedRte->subquery, recursivePlanningContext))
+		{
+			/*
+			 * RecursivelyPlanSubquery fails to plan a subquery only if it
+			 * contains references to the outer query. This means that, we can't
+			 * plan such outer joins (like <recurring LEFT OUTER distributed>)
+			 * if it's a LATERAL join where the distributed side is a subquery that
+			 * references the outer side, as in,
+			 *
+			 * SELECT * FROM reference
+			 * LEFT JOIN LATERAL
+			 * (SELECT * FROM distributed WHERE reference.b > distributed.b) q
+			 * USING (a);
+			 *
+			 * For this reason, we don't need to check such cases in
+			 * DeferredErrorIfUnsupportedRecurringTuplesJoin but instead
+			 * we capture such joins here and throw an error based on this
+			 * fact.
+			 */
+			Assert(ContainsReferencesToOuterQuery(distributedRte->subquery));
+
+			ereport(ERROR, (errmsg("cannot recursively plan the distributed "
+								   "subquery involved in an outer join since "
+								   "it contains reference(s) to outer query")));
+		}
+	}
+	else
+	{
+		/*
+		 * We don't expect RecursivelyPlanRecurringTupleOuterJoinWalker to try recursively
+		 * plan such an RTE.
+		 */
+		ereport(ERROR, errmsg("got unexpected RTE type (%d) when recursively "
+							  "planning a join",
+							  distributedRte->rtekind));
+	}
+}
+
+
+/*
+ * IsRTERefRecurring return true if given rte reference points to a recurring
```

Comment:
minor typo(s): 
- return->returns
- if given -> if the given
- safe assume -> safe to assume ?


Also, I _think_ the example subquery q should probably be in RecursivelyPlanRecurringTupleOuterJoinWalker where we call if (IsRTERefRecurring())..?

Reply:
* return->returns (fixed)
* if given -> if the given (indeded this sounds correct ? 😄)
* safe assume -> safe to assume ? (fixed)

> Also, I think the example subquery q should probably be in RecursivelyPlanRecurringTupleOuterJoinWalker where we call if (IsRTERefRecurring())..?

Couldn't follow what this exactly means